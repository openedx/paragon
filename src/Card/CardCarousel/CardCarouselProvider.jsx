import React, { createContext, useContext, useMemo } from 'react';
import PropTypes from 'prop-types';
import { OverflowScrollContext } from '../../OverflowScroll';

export const CardCarouselContext = createContext();

function CardCarouselProvider({
  columnSizes,
  hasInteractiveChildren,
  canScrollHorizontal,
  children,
  CardCarouselControls,
}) {
  const {
    overflowRef,
    setOverflowRef,
    isScrolledToStart,
    isScrolledToEnd,
    scrollToPrevious,
    scrollToNext,
  } = useContext(OverflowScrollContext);

  const cardCarouselContextValue = useMemo(() => ({
    overflowRef,
    setOverflowRef,
    columnSizes,
    hasInteractiveChildren,
    canScrollHorizontal,
    isScrolledToStart,
    isScrolledToEnd,
    scrollToPrevious,
    scrollToNext,
    CardCarouselControls,
  }), [
    overflowRef,
    setOverflowRef,
    columnSizes,
    hasInteractiveChildren,
    canScrollHorizontal,
    isScrolledToStart,
    isScrolledToEnd,
    scrollToPrevious,
    scrollToNext,
    CardCarouselControls,
  ]);

  return (
    <CardCarouselContext.Provider value={cardCarouselContextValue}>
      {children}
    </CardCarouselContext.Provider>
  );
}

CardCarouselProvider.propTypes = {
  children: PropTypes.node.isRequired,
  /**
   * An object containing the desired column size at each breakpoint, following a similar
   * props API as ``react-bootstrap/Col``
   */
  columnSizes: PropTypes.shape({
    xs: PropTypes.number,
    sm: PropTypes.number,
    md: PropTypes.number,
    lg: PropTypes.number,
    xl: PropTypes.number,
  }),
  /** Whether the child `Card` components are interactive/focusable. If not, a `tabindex="0"` is
   * added to be a11y-compliant */
  hasInteractiveChildren: PropTypes.bool,
  /** Whether the `CardDeck` supports horizontal scrolling when there are overflow children */
  canScrollHorizontal: PropTypes.bool,
  /** Optional override for the default `CardCarouselControls` */
  CardCarouselControls: PropTypes.elementType,
};

CardCarouselProvider.defaultProps = {
  columnSizes: {
    sm: 12,
    lg: 6,
    xl: 4,
  },
  hasInteractiveChildren: false,
  canScrollHorizontal: true,
  CardCarouselControls: undefined,
};

export default CardCarouselProvider;
