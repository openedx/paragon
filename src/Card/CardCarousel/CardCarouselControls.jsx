import React, { useContext } from 'react';
import { useIntl } from 'react-intl';
import Stack from '../../Stack';
import IconButton from '../../IconButton';
import Icon from '../../Icon';
import { ArrowForward, ArrowBack } from '../../../icons';
import { CardCarouselContext } from './CardCarouselProvider';
import messages from './messages';

function CardCarouselControls() {
  const {
    isScrolledToStart,
    isScrolledToEnd,
    scrollToPrevious,
    scrollToNext,
  } = useContext(CardCarouselContext);

  const intl = useIntl();

  if (isScrolledToStart && isScrolledToEnd) {
    return null;
  }

  return (
    <div className="pgn__card-carousel-controls">
      <Stack direction="horizontal" gap={1}>
        <IconButton
          disabled={isScrolledToStart}
          src={ArrowBack}
          iconAs={Icon}
          alt={intl.formatMessage(messages.scrollToPrevious)}
          onClick={scrollToPrevious}
        />
        <IconButton
          disabled={isScrolledToEnd}
          src={ArrowForward}
          iconAs={Icon}
          alt={intl.formatMessage(messages.scrollToNext)}
          onClick={scrollToNext}
        />
      </Stack>
    </div>
  );
}

export default CardCarouselControls;
