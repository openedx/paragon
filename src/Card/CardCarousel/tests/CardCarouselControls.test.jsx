import React from 'react';
import PropTypes from 'prop-types';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import renderer from 'react-test-renderer';
import { IntlProvider } from 'react-intl';
import { CardCarouselContext } from '../CardCarouselProvider';
import CardCarouselControls from '../CardCarouselControls';

const mockScrollToPrevious = jest.fn();
const mockScrollToNext = jest.fn();
const defaultCardCarouselContextValue = {
  isScrolledToStart: false,
  isScrolledToEnd: false,
  scrollToPrevious: mockScrollToPrevious,
  scrollToNext: mockScrollToNext,
};

function CardCarouselControlsWrapper({
  cardCarouselContextValue,
  ...rest
}) {
  return (
    <IntlProvider locale="en">
      <CardCarouselContext.Provider value={cardCarouselContextValue}>
        <CardCarouselControls {...rest} />
      </CardCarouselContext.Provider>
    </IntlProvider>
  );
}

CardCarouselControlsWrapper.propTypes = {
  cardCarouselContextValue: PropTypes.shape(),
};

CardCarouselControlsWrapper.defaultProps = {
  cardCarouselContextValue: defaultCardCarouselContextValue,
};

describe('<CardCarouselControls />', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('renders with disabled previous control when scrolled to start', () => {
    const contextValue = {
      ...defaultCardCarouselContextValue,
      isScrolledToStart: true,
    };
    const tree = renderer.create((
      <CardCarouselControlsWrapper cardCarouselContextValue={contextValue} />
    )).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders with disabled next control when scrolled to end', () => {
    const contextValue = {
      ...defaultCardCarouselContextValue,
      isScrolledToEnd: true,
    };
    const tree = renderer.create((
      <CardCarouselControlsWrapper cardCarouselContextValue={contextValue} />
    )).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders with enabled controls when not scrolled to start/end', () => {
    const contextValue = {
      ...defaultCardCarouselContextValue,
      isScrolledToStart: false,
      isScrolledToEnd: false,
    };
    const tree = renderer.create((
      <CardCarouselControlsWrapper cardCarouselContextValue={contextValue} />
    )).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders without controls when both previous/next controls would be disabled', () => {
    const contextValue = {
      ...defaultCardCarouselContextValue,
      isScrolledToStart: true,
      isScrolledToEnd: true,
    };
    const tree = renderer.create((
      <CardCarouselControlsWrapper cardCarouselContextValue={contextValue} />
    )).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('handles scroll to previous click', () => {
    const contextValue = {
      ...defaultCardCarouselContextValue,
      isScrolledToStart: false,
    };
    render((
      <CardCarouselControlsWrapper cardCarouselContextValue={contextValue} />
    ));
    userEvent.click(screen.getByLabelText('Scroll to previous'));
    expect(mockScrollToPrevious).toHaveBeenCalledTimes(1);
  });

  it('handles scroll to next click', () => {
    const contextValue = {
      ...defaultCardCarouselContextValue,
      isScrolledToEnd: false,
    };
    render((
      <CardCarouselControlsWrapper cardCarouselContextValue={contextValue} />
    ));
    userEvent.click(screen.getByLabelText('Scroll to next'));
    expect(mockScrollToNext).toHaveBeenCalledTimes(1);
  });
});
