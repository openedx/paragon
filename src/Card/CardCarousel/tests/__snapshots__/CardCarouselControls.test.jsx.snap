// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CardCarouselControls /> renders with disabled next control when scrolled to end 1`] = `
<div
  className="pgn__card-carousel-controls"
>
  <div
    className="pgn__hstack pgn__stack-gap--1"
  >
    <button
      aria-label="Scroll to previous"
      className="btn-icon btn-icon-primary btn-icon-md"
      disabled={false}
      onClick={[MockFunction]}
      type="button"
    >
      <span
        className="btn-icon__icon-container"
      >
        <span
          className="pgn__icon btn-icon__icon"
        >
          <svg
            aria-hidden={true}
            fill="none"
            focusable={false}
            height={24}
            role="img"
            viewBox="0 0 24 24"
            width={24}
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M20 11H7.83l5.59-5.59L12 4l-8 8 8 8 1.41-1.41L7.83 13H20v-2z"
              fill="currentColor"
            />
          </svg>
        </span>
      </span>
    </button>
    <button
      aria-label="Scroll to next"
      className="btn-icon btn-icon-primary btn-icon-md"
      disabled={true}
      onClick={[MockFunction]}
      type="button"
    >
      <span
        className="btn-icon__icon-container"
      >
        <span
          className="pgn__icon btn-icon__icon"
        >
          <svg
            aria-hidden={true}
            fill="none"
            focusable={false}
            height={24}
            role="img"
            viewBox="0 0 24 24"
            width={24}
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M12 4l-1.41 1.41L16.17 11H4v2h12.17l-5.58 5.59L12 20l8-8-8-8z"
              fill="currentColor"
            />
          </svg>
        </span>
      </span>
    </button>
  </div>
</div>
`;

exports[`<CardCarouselControls /> renders with disabled previous control when scrolled to start 1`] = `
<div
  className="pgn__card-carousel-controls"
>
  <div
    className="pgn__hstack pgn__stack-gap--1"
  >
    <button
      aria-label="Scroll to previous"
      className="btn-icon btn-icon-primary btn-icon-md"
      disabled={true}
      onClick={[MockFunction]}
      type="button"
    >
      <span
        className="btn-icon__icon-container"
      >
        <span
          className="pgn__icon btn-icon__icon"
        >
          <svg
            aria-hidden={true}
            fill="none"
            focusable={false}
            height={24}
            role="img"
            viewBox="0 0 24 24"
            width={24}
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M20 11H7.83l5.59-5.59L12 4l-8 8 8 8 1.41-1.41L7.83 13H20v-2z"
              fill="currentColor"
            />
          </svg>
        </span>
      </span>
    </button>
    <button
      aria-label="Scroll to next"
      className="btn-icon btn-icon-primary btn-icon-md"
      disabled={false}
      onClick={[MockFunction]}
      type="button"
    >
      <span
        className="btn-icon__icon-container"
      >
        <span
          className="pgn__icon btn-icon__icon"
        >
          <svg
            aria-hidden={true}
            fill="none"
            focusable={false}
            height={24}
            role="img"
            viewBox="0 0 24 24"
            width={24}
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M12 4l-1.41 1.41L16.17 11H4v2h12.17l-5.58 5.59L12 20l8-8-8-8z"
              fill="currentColor"
            />
          </svg>
        </span>
      </span>
    </button>
  </div>
</div>
`;

exports[`<CardCarouselControls /> renders with enabled controls when not scrolled to start/end 1`] = `
<div
  className="pgn__card-carousel-controls"
>
  <div
    className="pgn__hstack pgn__stack-gap--1"
  >
    <button
      aria-label="Scroll to previous"
      className="btn-icon btn-icon-primary btn-icon-md"
      disabled={false}
      onClick={[MockFunction]}
      type="button"
    >
      <span
        className="btn-icon__icon-container"
      >
        <span
          className="pgn__icon btn-icon__icon"
        >
          <svg
            aria-hidden={true}
            fill="none"
            focusable={false}
            height={24}
            role="img"
            viewBox="0 0 24 24"
            width={24}
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M20 11H7.83l5.59-5.59L12 4l-8 8 8 8 1.41-1.41L7.83 13H20v-2z"
              fill="currentColor"
            />
          </svg>
        </span>
      </span>
    </button>
    <button
      aria-label="Scroll to next"
      className="btn-icon btn-icon-primary btn-icon-md"
      disabled={false}
      onClick={[MockFunction]}
      type="button"
    >
      <span
        className="btn-icon__icon-container"
      >
        <span
          className="pgn__icon btn-icon__icon"
        >
          <svg
            aria-hidden={true}
            fill="none"
            focusable={false}
            height={24}
            role="img"
            viewBox="0 0 24 24"
            width={24}
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M12 4l-1.41 1.41L16.17 11H4v2h12.17l-5.58 5.59L12 20l8-8-8-8z"
              fill="currentColor"
            />
          </svg>
        </span>
      </span>
    </button>
  </div>
</div>
`;

exports[`<CardCarouselControls /> renders without controls when both previous/next controls would be disabled 1`] = `null`;
